### PR TITLE
Introducing Swagger in Dredd

### PR DESCRIPTION
**Pull request by honzajavorek**
Also addressing:

- a bit of https://github.com/apiaryio/dredd/issues/353
- closes https://github.com/apiaryio/dredd/issues/496
- closes https://github.com/apiaryio/dredd/issues/389

### CODE REVIEW

File: docs/how-it-works.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+1. `+ Schema` section - provided custom JSON Schema (draft v4 or v3) will be used.
+2. `+ Attributes` section with data structure description in [MSON][] - API Blueprint parser automatically generates JSON Schema from MSON.
+3. `+ Body` section with sample JSON payload - [Gavel.js][], which is responsible for validation in Dredd, automatically infers some basic expectations described below.
+
+This order [exactly follows the API Blueprint specification](https://github.com/apiaryio/api-blueprint/blob/master/API%20Blueprint%20Specification.md#relation-of-body-schema-and-attributes-sections).
+
+#### Swagger
+
+1. [`response.schema`](http://swagger.io/specification/#responseSchema) - provided JSON Schema will be used.
```
Comment (pksunkara) [inline docs/how-it-works.md]: I think response.examples is given preference.

Comment (honzajavorek) [inline docs/how-it-works.md]: when checking validity of response? I think it shouldn't

File: README.md [hunk 2]
```diff
@@ -20,6 +19,11 @@ Dredd reads your API description and step by step validates whether your API
 implementation replies with responses as they are described in the
 documentation.
 
+## API Description Formats Support
```
Comment (netmilk) [inline README.md]: Maybe "Supported API Description formats" ?

File: docs/example.md [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 This is an example how to create a simple [Express.js][] API backend application tested with Dredd
 
-## Create API Description File
+## Create an API Description Document
```
Comment (netmilk) [inline docs/example.md]: Is the article correct?

File: docs/how-it-works.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+## Limitations of API Description Formats
+
+## Automatic Expectations
+
+Dredd automatically generates expectations on HTTP responses based on examples in the API description with use of [Gavel.js][] library. Please refer to [Gavel][] rules if you want know more.
+
+### Response Headers Expectations
+
+- All headers specified in the API description must be present in the response.
+- Names of headers are validated in case-insensitive way.
```
Comment (netmilk) [inline docs/how-it-works.md]: Maybe "validated >the< case-insensitive way" ?

File: docs/how-it-works.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+- Only values of headers significant for content negotiation are validated.
+- All other headers values can differ.
+
+When using [Swagger][], headers are taken from [`response.headers`][]. HTTP headers significant for content negotiation are inferred according to following rules:
+
+- [`produces`][] is propagated as response's `Content-Type` header.
+- Response's `Content-Type` header overrides any `produces`.
```
Comment (netmilk) [inline docs/how-it-works.md]: This is how it should be, but it is buggy in the implementation, right?

File: docs/how-it-works.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+### Response Body Expectations
+
+By default Dredd validates HTTP response body against [JSON Schema][] inferred from the API description under test. The effective JSON Schema is taken from following places (the order goes from the highest priority to the lowest):
```
Comment (netmilk) [inline docs/how-it-works.md]: I would rephrase this to say - by default it validates only structure of the JSON bodies

JSON schema used on background in Gavel is something what's not important for the user in this paragraph.

File: docs/how-it-works.md (hunk 5 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: yes, should be fixed or filed as an issue on Dredd/Swagger Adapter cc @pksunkara

File: docs/how-it-works.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+> **Note:** The implementation of API Blueprint's request-specific parameters is still in progress and there's only experimental support for it in Dredd as of now.
+
+### Request Headers
+
+In [Swagger][] documents, HTTP headers are inferred from [`"in": "header"` parameters][parameters]. HTTP headers significant for content negotiation are inferred according to following rules:
+
+- [`consumes`][] is propagated as request's `Content-Type` header.
+- [`produces`][] is propagated as request's `Accept` header.
+
```
Comment (netmilk) [inline docs/how-it-works.md]: I'd mention here inferred content-type header can be overridden by explicitly setting the in header parameter. And it's a known bug, because it doesn't work like that atm.

File: docs/how-it-works.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+> **Note:** The implementation of API Blueprint's request-specific parameters is still in progress and there's only experimental support for it in Dredd as of now.
+
+### Request Headers
+
+In [Swagger][] documents, HTTP headers are inferred from [`"in": "header"` parameters][parameters]. HTTP headers significant for content negotiation are inferred according to following rules:
+
+- [`consumes`][] is propagated as request's `Content-Type` header.
+- [`produces`][] is propagated as request's `Accept` header.
+- If request body parameters are specified as `"in": "form"`, request's `Content-Type` header is set to `application/x-www-form-urlencoded`.
```
Comment (honzajavorek) [inline docs/how-it-works.md]: Also I'm not sure this is 100% true as of now. Should be fixed or filed as an issue on Dredd/Swagger Adapter cc @pksunkara @netmilk

File: docs/how-it-works.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+1. [`+ Body`][] section with sample JSON payload.
+2. [`+ Attributes`][] section with data structure description in [MSON][] - API Blueprint parser automatically generates sample JSON payload from MSON.
+
+This order [exactly follows the API Blueprint specification][body-schema-attributes].
+
+#### Swagger
+
+The effective request body is inferred from [`"in": "body"` and `"in": "form"` parameters][parameters]. Since Swagger doesn't provide any way to specify raw JSON sample for the request body, Dredd's [Swagger Adapter][] generates some sample values from the JSON Schema provided in the [`schema`][] property. Following rules apply when the adapter fills values of the properties, ordered by precedence:
+
+1. Value of `default`.
```
Comment (netmilk) [inline docs/how-it-works.md]: What about [schema examples](http://swagger.io/specification/#schemaExample)? Please verify how it works and clarify it here.

File: docs/how-it-works.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+> **Note:** Improving the support for multiple requests and responses is under development. Refer to issues [#25](https://github.com/apiaryio/dredd/issues/25) and [#78](https://github.com/apiaryio/dredd/issues/78) for details. Support for URI parameters specific to a single request within one action is also limited. Solving [#227](https://github.com/apiaryio/dredd/issues/227) should unblock many related problems. Also see [Multiple Requests and Responses within One API Blueprint Action](how-to-guides.md#multiple-requests-and-responses-within-one-api-blueprint-action) guide for workarounds.
+
+#### Swagger
+
+The [Swagger][] format allows to specify multiple responses for a single operation.
+By default Dredd tests only responses with `2xx` status codes. Responses with other
+codes are marked as _skipped_ and can be activated in [hooks](hooks.md) - see [Testing non-2xx Responses with Swagger](how-to-guides.md#testing-non-2xx-responses-with-swagger).
+
+[Default responses][default responses] are ignored by Dredd. Also, as of now,
+only `application/json` media type is supported in [`produces`][] and [`consumes`][].
```
Comment (netmilk) [inline docs/how-it-works.md]: This should be extended for `whatever+json` media types and for support of `application/json; whatever`. In other words, to support JSON suffixed media types and omit the parameters.

Consider using this npm: https://www.npmjs.com/package/media-typer

File: docs/index.md [hunk 11]
```diff
@@ -9,13 +9,17 @@
...
 > **Dredd is a language-agnostic command-line tool for validating
-API description written in [API Blueprint][] format against its backend
-implementation.**
+API description document against backend implementation of the API.**
 
 Dredd reads your API description and step by step validates whether your API
 implementation replies with responses as they are described in the
 documentation.
 
+## API Description Formats Support
```
Comment (netmilk) [inline docs/index.md]: Maybe: "Supported API Description Formats" as well?

File: docs/how-it-works.md (hunk 5 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: https://github.com/apiaryio/fury-adapter-swagger/issues/65

File: README.md (hunk 2 above)
Comment (honzajavorek) [inline README.md]: Addressed in 5e5c81bdf4a78b98edf4172e580823ffd6308686.

File: docs/example.md (hunk 3 above)
Comment (honzajavorek) [inline docs/example.md]: Addressed in 5e5c81bdf4a78b98edf4172e580823ffd6308686.

File: docs/how-it-works.md (hunk 4 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: Addressed in 5e5c81bdf4a78b98edf4172e580823ffd6308686.

File: docs/how-it-works.md (hunk 6 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: Addressed in 5e5c81bdf4a78b98edf4172e580823ffd6308686.

File: docs/how-it-works.md (hunk 10 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: I think this is a larger story - AFAIK most of the Dredd stack can't deal with `+json` media types (namely Gavel or see this PR https://github.com/apiaryio/dredd/pull/262).

File: docs/how-it-works.md (hunk 8 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: https://github.com/apiaryio/fury-adapter-swagger/issues/66

File: docs/how-it-works.md (hunk 10 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: https://github.com/apiaryio/dredd/issues/531

File: docs/how-it-works.md (hunk 9 above)
Comment (honzajavorek) [inline docs/how-it-works.md]: Checked with Pavan and addressed in 1ff8e52.

**Comment by honzajavorek** [comment]
@netmilk All comments addressed. Good to merge?

**Comment by honzajavorek** [comment]
:heart:
